### PR TITLE
Cache available tier names on DiscoveryNode and DiscoveryNodes

### PR DESCRIPTION
**Pull request by original-brownbear**
We burn almost 2% of the runtime in many-shards benchmark bootstrapping on looping
through the roles just to compare names for the data tier allocation decider.
This change makes that step a lot cheaper by caching the role names in sets
that only need to be build rarely.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 1]
```diff
@@ -377,6 +389,17 @@ public DiscoveryNode(StreamInput in) throws IOException {
         } else {
             this.externalId = nodeName;
         }
+        this.roleNames = Set.of(roleNames);
+    }
+
+    /**
+     * Check if node has the role with the given {@code roleName}.
+     *
+     * @param roleName role nae to check
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:398]:
```suggestion
     * @param roleName role name to check
```

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 2]
```diff
@@ -111,6 +111,8 @@ public static boolean isDedicatedFrozenNode(final Settings settings) {
     private final Map<String, String> attributes;
     private final Version version;
     private final SortedSet<DiscoveryNodeRole> roles;
+
+    private final Set<String> roleNames;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:115]: Yea it's used in a couple of places still currently.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2

**Comment by original-brownbear** [comment]
@idegtiarenko ping in case you have a second to give this another look :)

**Review by idegtiarenko — APPROVED** [review]

**Pull request merged**